src/uct/ib/ud/ud_inl.h

```diff
@@ -66,25 +66,38 @@ static inline uct_ud_send_skb_t *uct_ud_iface_get_tx_skb(uct_ud_iface_t *iface,
         return NULL;
     }
 
-    if (ep->tx.psn == ep->tx.max_psn) {
+    if (ucs_unlikely(ep->tx.psn == ep->tx.max_psn)) {
         ucs_trace_data("iface=%p ep=%p (%d->%d) tx window full (max_psn=%u)",
                        iface, ep, ep->ep_id, ep->dest_ep_id, (unsigned)ep->tx.max_psn);
         return NULL;
     }
 
-    skb = ucs_mpool_get(iface->tx.mp);
-    if (!skb) {
+    skb = iface->tx.skb;
+    if (ucs_unlikely(!skb)) {
         ucs_trace_data("iface=%p out of tx skbs", iface);
         return NULL;
     }
+    ucs_prefetch(skb->neth);
     return skb;
 }
 
+/* same as function above but also check that iface is connected */
+static inline uct_ud_send_skb_t *uct_ud_iface_get_tx_skb2(uct_ud_iface_t *iface,
+                                                          uct_ud_ep_t *ep)
+{
+    if (ucs_unlikely(!uct_ud_ep_is_connected(ep))) {
```

Comment:
also, need to avoid this additional check, instead make the window=0 before ep is connected